pkg/foundation/semaphore/semaphore.go

```diff
@@ -16,71 +16,102 @@ package semaphore
 
 import (
 	"sync"
-
-	"github.com/conduitio/conduit/pkg/foundation/cerrors"
 )
 
-// Simple provides a way to bound concurrent access to a resource. It only
-// allows one caller to gain access at a time.
+// Simple provides a way to lock concurrent access to a resource. It only allows
+// one caller to gain access at a time.
 type Simple struct {
-	// lastTicket holds the last issued ticket.
-	lastTicket Ticket
-	// mu guards concurrent access to lastTicket.
-	mu sync.Mutex
+	// last holds the last issued lock.
+	last Lock
+
+	chanPool sync.Pool
+	init     sync.Once
 }
 
 // Ticket reserves a place in the queue and can be used to acquire access to a
-// resource.
+// Lock.
 type Ticket struct {
-	// ready is closed when the ticket acquired the semaphore.
-	ready chan struct{}
-	// next is closed when the ticket is released.
-	next chan struct{}
+	Lock
+	// ready is the same channel as Lock.next of the previous lock. Once the
+	// previous lock is released, the value i will be sent into ready, signaling
+	// this ticket that it can acquire a Lock.
+	ready chan int
+}
+
+// Lock represents a lock of the semaphore. After the lock is not needed any
+// more it should be released.
+type Lock struct {
+	// i is the index of the current lock.
+	i int
+	// next is the same channel as Ticket.ready of the next ticket. Once Lock is
+	// released, the value i+1 will be sent into next, signaling the next ticket
+	// that it can acquire a Lock.
+	next chan int
 }
 
 // Enqueue reserves the next place in the queue and returns a Ticket used to
-// acquire access to the resource when it's the callers turn. The Ticket has to
-// be supplied to Release before discarding.
+// acquire access to the Lock when it's the callers turn. The Ticket has to be
+// supplied to Acquire and discarded afterwards.
+// Enqueue is not safe for concurrent use.
 func (s *Simple) Enqueue() Ticket {
-	s.mu.Lock()
-	defer s.mu.Unlock()
+	s.init.Do(func() {
+		s.chanPool.New = func() any {
+			return make(chan int, 1) // make it buffered to not block the release of a lock
+		}
+
+		s.last.next = s.chanPool.Get().(chan int)
+		s.last.next <- s.last.i + 1 // first lock can be acquired right away
```

Comment:
Since this is done only during initialization, we should be able to write `s.last.next <- 1`? Or is something else possible due to which we need `s.last.i + 1`?